### ISSUE TITLE
`id` of publicKey is the same as the subject `id`

### ISSUE THREAD

**Issue by awoie** [digitalbazaar/did-method-key-js#6]
In the following example DID Document:

```
{
  "@context": "https://w3id.org/did/v1",
  "id": "did:key:z6MknxumoAwcGrjM3M4N8bWSTDZLwRgMxChytWqqWe2E4d14",
  "assertionMethod": [
    "did:key:z6MknxumoAwcGrjM3M4N8bWSTDZLwRgMxChytWqqWe2E4d14"
  ],
  "authentication": [
    "did:key:z6MknxumoAwcGrjM3M4N8bWSTDZLwRgMxChytWqqWe2E4d14"
  ],
  "capabilityDelegation": [
    "did:key:z6MknxumoAwcGrjM3M4N8bWSTDZLwRgMxChytWqqWe2E4d14"
  ],
  "capabilityInvocation": [
    "did:key:z6MknxumoAwcGrjM3M4N8bWSTDZLwRgMxChytWqqWe2E4d14"
  ],
  "keyAgreement": [
    {
      "id": "did:key:z6MknxumoAwcGrjM3M4N8bWSTDZLwRgMxChytWqqWe2E4d14#zC3hrJ5iJGYJ8FmaChkJf66ttRGkMh6gtpZuqM2hc9dEcK",
      "type": "X25519KeyAgreementKey2019",
      "controller": "did:key:z6MknxumoAwcGrjM3M4N8bWSTDZLwRgMxChytWqqWe2E4d14",
      "publicKeyBase58": "4D8aiXV5AYqpfXEFLp67wjD82UYjXzHKg2YKBcokWzW7"
    }
  ],
  "publicKey": [
    {
      "id": "did:key:z6MknxumoAwcGrjM3M4N8bWSTDZLwRgMxChytWqqWe2E4d14",
      "type": "Ed25519VerificationKey2018",
      "controller": "did:key:z6MknxumoAwcGrjM3M4N8bWSTDZLwRgMxChytWqqWe2E4d14",
      "publicKeyBase58": "9WejCvhAwKEsvrDfT2Ybc81M7rQWYKTdCVvugN4D9QDg"
    }
  ]
}
```

... the `id` of the public `Ed25519VerificationKey2018` is the same as the `id` of the subject of the DID Document. Isn't the idea that `id` values should be unique and refer to one specific thing?

**Comment by awoie** [comment]
@dmitrizagidulin Am I misinterpreting the DID spec?

**Comment by dlongley** [comment]
@awoie,

We're changing this, it's a mistake. I do want to say that this would not necessarily have been illegal, as the subject would be understood to be a key itself. However, that's not what we want for `did:key` so we're fixing it. Thanks!

**Comment by awoie** [comment]
We see a lot of DID URLs to link to public keys in DID Documents, e.g., `did:ex:0xab#asdf-key1`. 

From the DID resolution spec [https://w3c-ccg.github.io/did-resolution/#dereferencing-algorithm]:
> If present, separate the DID Fragment from the input DID URL. Execute the algorithm for § 4.3 Dereferencing the Primary Resource, with the input DID URL adjusted accordingly.

In the above, I believe "Dereferencing the Primary Resource" has to be replaced with "Dereferencing the Secondary Resource".

From the spec [https://w3c-ccg.github.io/did-resolution/#dereferencing-algorithm-secondary]
> From the resolved DID Document, select the JSON-LD object whose id property matches the input DID URL, e.g. a public key or service endpoint in the DID Document.

@dlongley I'm wondering how this could be achieved with the approach you have taken here. Could you provide an example of how this would work? Wouldn't this result in an infinite recursion?

**Comment by awoie** [comment]
@dlongley I double-checked the spec and as far as I understood it won't resolve in an infinite loop. Having the `id` as the subject of the DID Doc and as the `id` of the key would just render DID URLs to that specific key impossible. Is this correct?

**Comment by OR13** [comment]
I had some issues with this previously, I will the didDoc looked more like this:

```
{
      "@context": "https://w3id.org/did/v1",
      "id": "did:key:z6MkobPDEBzDVqEi9VanhLMPM3ZzuumvYDrSyjG4ULZ57pcK",
      "publicKey": [
        {
          "id": "did:key:z6MkobPDEBzDVqEi9VanhLMPM3ZzuumvYDrSyjG4ULZ57pcK",
          "type": "Ed25519VerificationKey2018",
          "controller": "did:key:z6MkobPDEBzDVqEi9VanhLMPM3ZzuumvYDrSyjG4ULZ57pcK",
          "publicKeyBase58": "A98AdwjnAHkF2zk61mPYVx216LW58Lc6HiM8e4b4Cbpw"
        }
      ],
      "authentication": [
        "did:key:z6MkobPDEBzDVqEi9VanhLMPM3ZzuumvYDrSyjG4ULZ57pcK#8R_gUPjBoJ_nf39_G7VLGWNuhL5etuW7zvS46kwYN6Q"
      ],
      "assertionMethod": [
        "did:key:z6MkobPDEBzDVqEi9VanhLMPM3ZzuumvYDrSyjG4ULZ57pcK#8R_gUPjBoJ_nf39_G7VLGWNuhL5etuW7zvS46kwYN6Q"
      ],
      "capabilityDelegation": [
        "did:key:z6MkobPDEBzDVqEi9VanhLMPM3ZzuumvYDrSyjG4ULZ57pcK#8R_gUPjBoJ_nf39_G7VLGWNuhL5etuW7zvS46kwYN6Q"
      ],
      "capabilityInvocation": [
        "did:key:z6MkobPDEBzDVqEi9VanhLMPM3ZzuumvYDrSyjG4ULZ57pcK#8R_gUPjBoJ_nf39_G7VLGWNuhL5etuW7zvS46kwYN6Q"
      ],
      "keyAgreement": [
        {
          "id": "did:key:z6MkobPDEBzDVqEi9VanhLMPM3ZzuumvYDrSyjG4ULZ57pcK#zBzdKZByVW5acSrodNdfbSBowCiiXFGfDozz6VH4TNfzvq",
          "type": "X25519KeyAgreementKey2019",
          "controller": "did:key:z6MkobPDEBzDVqEi9VanhLMPM3ZzuumvYDrSyjG4ULZ57pcK",
          "publicKeyBase58": "8bqpngJhqL1kkevEB2U2eFuUSiHhxcK76oTRyeyZkpd"
        }
      ]
    }
```

**Comment by msporny** [comment]
@awoie @OR13 -- it's a bug, it's wrong, it needs to be fixed. @dmitrizagidulin has the action to fix it.

**Comment by dmitrizagidulin** [comment]
Wait, I think I'm missing something - was this issue not fixed in https://github.com/digitalbazaar/did-method-key-js/pull/4 ?

**Comment by dmitrizagidulin** [comment]
@awoie / @OR13 - I think this issue has been fixed in a previous commit; is it ok if I close this?

**Comment by OR13** [comment]
@dmitrizagidulin I would note the npm version of the fix before closing. latest (0.2.0) appears to still have this issue.

**Comment by dmitrizagidulin** [comment]
@OR13 ah, whoops, you're right, that latest fix wasn't published to npm. It should be published now, v0.3.0!

**Comment by OR13** [comment]
Closing the loop for us on this, we deployed a new version that implements these changes here: https://github.com/transmute-industries/did-key/commit/491bbcdee2deb7cbb475ba7a0f43dc647c36bea1
